pkg/services/hivelocity/hvtag/hvtag.go

```diff
@@ -0,0 +1,176 @@
+func DeviceTagFromList(key DeviceTagKey, tagList []string) (DeviceTag, error) {
+	var found bool
+	var deviceTag DeviceTag
+	var err error
+
+	for _, tagString := range tagList {
+		// filter out irrelevant tagStrings quickly
+		if !strings.HasPrefix(tagString, key.Prefix()) {
+			continue
+		}
+
+		// get DeviceTag from tagString
+		deviceTag, err = deviceTagFromString(tagString)
+		if err != nil {
+			continue
+		}
+
+		// additional check if key is correct - probably not necessary due to HasPrefix check
+		if deviceTag.Key != key {
+			continue
+		}
+
+		// Check whether a correct DeviceTag has been found already. If so, return with error.
+		if found {
+			return DeviceTag{}, ErrMultipleDeviceTagsFound
+		}
+		found = true
+	}
+
+	if !found {
+		return DeviceTag{}, ErrDeviceTagNotFound
+	}
+
+	return deviceTag, nil
+}
+
+// MachineTagFromList returns the machine tag from a list of tag strings.
+func MachineTagFromList(tagList []string) (DeviceTag, error) {
+	return DeviceTagFromList(DeviceTagKeyMachine, tagList)
+}
+
+// ClusterTagFromList returns the machine tag from a list of tag strings.
+func ClusterTagFromList(tagList []string) (DeviceTag, error) {
+	return DeviceTagFromList(DeviceTagKeyCluster, tagList)
+}
+
+// DeviceTypeTagFromList returns the machine tag from a list of tag strings.
+func DeviceTypeTagFromList(tagList []string) (DeviceTag, error) {
+	return DeviceTagFromList(DeviceTagKeyDeviceType, tagList)
+}
+
+// deviceTagFromString takes the tag of a HV device and returns a DeviceTag or an error if it is invalid.
+func deviceTagFromString(tagString string) (DeviceTag, error) {
+	tagElements := strings.Split(tagString, "=")
+	if len(tagElements) != 2 {
+		return DeviceTag{}, ErrDeviceTagInvalidFormat
+	}
+
+	key := DeviceTagKey(tagElements[0])
+	value := tagElements[1]
+
+	if key == "" {
+		return DeviceTag{}, ErrDeviceTagEmptyKey
+	}
+	if value == "" {
+		return DeviceTag{}, ErrDeviceTagEmptyValue
+	}
+
+	if !key.IsValid() {
+		return DeviceTag{}, ErrDeviceTagKeyInvalid
+	}
+
+	return DeviceTag{key, value}, nil
+}
+
+// ToString returns the string representation of a DeviceTag object.
+func (deviceTag DeviceTag) ToString() string {
+	return string(deviceTag.Key) + "=" + deviceTag.Value
+}
+
+// IsInStringList checks whether a DeviceTag object can be found in a list of tag strings.
+func (deviceTag DeviceTag) IsInStringList(tagList []string) bool {
+	return slices.Contains(tagList, deviceTag.ToString())
+}
+
+// RemoveFromList removes all tag strings from a list that equal the string representation of DeviceTag.
+func (deviceTag DeviceTag) RemoveFromList(tagList []string) (newTagList []string, updated bool) {
+	newTagList = make([]string, 0, len(tagList))
+	for _, tagString := range tagList {
+		// append all tag strings to newTagList which do not equal this device tag
+		if tagString == deviceTag.ToString() {
+			updated = true
+		} else {
+			newTagList = append(newTagList, deviceTag.ToString())
```

Comment:
I would use do `deviceTagString := deviceTag.ToString()` once before the loop, and then use the variable, and not call the method again and again. But in this context the performance difference should not matter much :-)

Reply:
not relevant any more. Only appears once